test/hotspot/jtreg/compiler/c2/irTests/RShiftLNodeIdealizationTests.java

```diff
@@ -89,4 +101,47 @@ public long test3(long x, long y) {
     public long test4(long x, long y) {
         return (x >> 62) > 4 ? 0L : 1L;
     }
+
+    final static int test5Shift = RunInfo.getRandom().nextInt(64);
+    final static long test5Mask = -1L << test5Shift;
+
+    @Test
+    @IR(counts = { IRNode.RSHIFT_L, "1" })
+    @IR(failOn = { IRNode.AND_L })
+    public long test5(long x) {
+        return (x & test5Mask) >> test5Shift;
+    }
+
+    final static int test6Shift = RunInfo.getRandom().nextInt(Integer.MAX_VALUE / 64) * 64 ;
+
+    @Test
+    @IR(failOn = { IRNode.RSHIFT_L })
+    public long test6(long x) {
+        return (x >> test6Shift);
+    }
+
+    final static int test7Shift = RunInfo.getRandom().nextInt(32) + 32;
+    final static long test7Min = -1L << (64 - test7Shift -1);
+    final static long test7Max = ~test7Min;
+
+    @Test
+    @IR(failOn = { IRNode.RSHIFT_L, IRNode.LSHIFT_L })
+    public long test7(long x) {
+        x = Integer.max(Integer.min((int)x, (int)test7Max), (int)test7Min);
+        return ((x << test7Shift) >> test7Shift);
+    }
+
+    @Test
+    @IR(counts = { IRNode.RSHIFT_L, "1", IRNode.LSHIFT_L, "1" })
+    public long test8(long x) {
+        x = Integer.max(Integer.min((int)x, (int)(test7Max+1)), (int)test7Min);
+        return ((x << test7Shift) >> test7Shift);
+    }
+
+    @Test
+    @IR(counts = { IRNode.RSHIFT_L, "1", IRNode.LSHIFT_L, "1" })
+    public long test9(long x) {
+        x = Integer.max(Integer.min((int)x, (int)test7Max), (int)(test7Min-1));
+        return ((x << test7Shift) >> test7Shift);
```

Comment:
It could be nice to have some test cases where both shift values are completely randomized.

Reply:
The transformation only happens if the amounts we shift left and right are the same. So if they are random, the transformation won't apply most of the time and, rarely, it will (because they will turn out to be the same). I'm not sure how to write an IR test then.

Reply:
You would not have to assert anything about the IR, just do value verification in that case.

Reply:
Added in new commit. Is that what you had in mind?